### PR TITLE
Use a REPL temporary directory when running an MLflow project on Databricks

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

This PR makes changes to use a REPL temporary directory when running an MLflow project on Databricks to ensure `mlflow.projects.run` works in a multi-use shared cluster.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

Manually tested on Databricks.

```python
import tempfile
import mlflow
from pathlib import Path

with tempfile.TemporaryDirectory() as tmpdir:
    tmpdir = Path(tmpdir)
    tmpdir.joinpath("python_env.yaml").write_text(
        """
python: "3.8.13"
build_dependencies:
  - pip
dependencies:
  - -r requirements.txt
"""
    )
    tmpdir.joinpath("requirements.txt").write_text(
        """
mlflow
scikit-learn
"""
    )
    tmpdir.joinpath("entrypoint.py").write_text(
        """
import sklearn

print(sklearn.__version__)
"""
    )

    tmpdir.joinpath("MLproject").write_text(
        """
name: virtualenv-example
python_env: python_env.yaml
entry_points:
  main:
    command: 'python entrypoint.py'
"""
    )
    mlflow.projects.run(str(tmpdir))
```

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/projects/backend/local.py [hunk 1]
```diff
@@ -142,8 +145,12 @@ def run(
...
                     else _PythonEnv()
                 )
-            python_bin_path = _install_python(python_env.python)
-            env_root = _get_mlflow_virtualenv_root()
+
+            repl_tmp_dir = Path("/tmp/repl_tmp_data", get_repl_id())
+            pyenv_root = repl_tmp_dir / _PYENV_ROOT_DIR
+            python_bin_path = _install_python(python_env.python, pyenv_root=pyenv_root)
+            # env_root = _get_mlflow_virtualenv_root()
+            env_root = repl_tmp_dir / _VIRTUALENV_ENVS_DIR
```
Comment (harupy) [inline mlflow/projects/backend/local.py:153]: Installing python in `/databricks/.pyenv` has the following issues: 

1. We need to modify the permission of `/databricks/.pyenv`.
2. If multiple users run `pyenv install` at the same time, that can cause an issue.
3. We need to modify the permission and ownership of the files downloaded/created by `pyenv install` after it completes to allow other users to access them.

File: mlflow/projects/backend/local.py [hunk 2]
```diff
@@ -142,11 +146,18 @@ def run(
                     if project.env_config_path
                     else _PythonEnv()
                 )
-            python_bin_path = _install_python(python_env.python)
-            env_root = _get_mlflow_virtualenv_root()
+
+            if is_in_databricks_runtime():
+                nfs_tmp_dir = get_or_create_nfs_tmp_dir()
```
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/projects/backend/local.py:151]: @WeichenXu123 Why should we use it?

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/projects/backend/local.py:151]: Does that return a directory under `/local_disk0`?

[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/projects/backend/local.py:151]: Using `get_or_create_nfs_tmp_dir` is fine. And it has benifit that we can reusing python env cache that created by spark udf with env-manager

[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/projects/backend/local.py:151]: We can reusing the pip cache that spark udf uses.
if you use the subdirectory.
 `{get_or_create_nfs_tmp_dir} / env / `

[the remark this answers is not in the document]
Comment (WeichenXu123) [reply mlflow/projects/backend/local.py:151]: spark udf pyenv pip cache is under
`{get_or_create_nfs_tmp_dir} / env /pip_cache_pkgs`

[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/projects/backend/local.py:152]: Updated

**Review by WeichenXu123 — APPROVED** [review]

File: mlflow/projects/backend/local.py [hunk 3]
```diff
@@ -142,12 +147,24 @@ def run(
...
+                envs_dir = Path(nfs_tmp_dir) / "envs"
+                pyenv_root = envs_dir / _PYENV_ROOT_DIR
+                virtualenv_root = envs_dir / _VIRTUALENV_ENVS_DIR
+                env_vars = _get_virtualenv_extra_env_vars(str(envs_dir))
+            else:
+                pyenv_root = None
+                virtualenv_root = Path(_get_mlflow_virtualenv_root())
+                env_vars = None
```
Comment (WeichenXu123) [inline mlflow/projects/backend/local.py:160]: Nit: The part is very similar to some code in `_get_or_create_virtualenv`, can we reuse that function ?

Comment (harupy) [reply mlflow/projects/backend/local.py:160]: Sounds good

**Comment by dbczumar** [comment]
@harupy Is this ready to merge?

**Comment by harupy** [comment]
@dbczumar Not yet, I need to push one more commit to address https://github.com/mlflow/mlflow/pull/7423#discussion_r1036668760.

File: mlflow/projects/backend/local.py (hunk 3 above)
Comment (harupy) [reply mlflow/projects/backend/local.py]: @WeichenXu123 I'll do refactoring in a follow-up PR.

**Comment by harupy** [comment]
Should be ready for merge now. I'll test `mlflow.projects.run` in a shared cluster.

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
Status of test ?

**Pull request merged**